include/rmm/detail/dynamic_load_runtime.hpp

```diff
@@ -110,17 +108,30 @@ struct dynamic_load_runtime {
  * This allows RMM users to compile/link against CUDA 11.2+ and run with
  * < CUDA 11.2 runtime as these functions are found at call time.
  */
+
+extern "C" {
+cudaError_t cudaMemPoolCreate(cudaMemPool_t*, const cudaMemPoolProps*) __attribute((weak));
+cudaError_t cudaMemPoolSetAttribute(cudaMemPool_t, cudaMemPoolAttr, void*) __attribute((weak));
+cudaError_t cudaMemPoolDestroy(cudaMemPool_t) __attribute((weak));
+cudaError_t cudaMallocFromPoolAsync(void**, size_t, cudaMemPool_t, cudaStream_t)
+  __attribute((weak));
+cudaError_t cudaFreeAsync(void*, cudaStream_t) __attribute((weak));
+cudaError_t cudaDeviceGetDefaultMemPool_sig(cudaMemPool_t*, int) __attribute((weak));
+}
+
```

Comment:
*question*: Why do we need to forward-declare this functions with weak symbols when previously we did not? Since this is gated behind CUDART >= 11020, I would have thought that `cuda_runtime_api.h` would be providing these forward declarations anyway?

Reply:
I think we this approach we can remove the `CUDART >= 11020` checks.

>  Why do we need to forward-declare this functions with weak symbols when previously we did not?

Because previously we only used these functions at compile time if we built statically and against a CUDART >= 11.2. Now we are saying that when statically or shared linking against 11.2+ we don't require the symbols at runtime, so that the loader doesn't hard error when < 11.2 is used. For example you could statically link against 11.1, and fallback to runtime loading 11.8 for these symbols now ( once I remove the above ifdef guard ).

Reply:
OK, so if I understand it, with these changes `RMM_CUDART_API_WRAPPER(some_name, some_signature)` needs to be able to resolve `some_name` at compile time (even if only as a weak symbol that will be overridden). Then, at runtime, assuming we linked against cudart dynamically rather than statically, the initial `static_cast<some_signature>(::some_name)` will return null, and we will try and dlsym then name and fail.

But, I don't understand why we need the forward declarations. If we're compiling with (shared) cudart >= 11.2, then `cuda_runtime_api.h` forward-declares the functions, so the names are defined. OK, if we compile with `--cudart none` then we get link errors because the cuda runtime header doesn't forward declare with `__attribute((weak))` but let me sweep that under the carpet.

Conversely, if we're compiling with cudart < 11.2, then the names are not defined by `cuda_runtime_api.h`, but then this code is ifdeffed away, so we don't need the foward declarations, because we never refer to the names anyway.

Reply:
The guards around the weak forward declares and `async_alloc` need to stay in place. The `async_alloc` and related async pool allocators aren't written in a way that would allow us to remove them (functions all use types found in 11.2+)

Reply:
> But, I don't understand why we need the forward declarations. If we're compiling with (shared) cudart >= 11.2, then cuda_runtime_api.h forward-declares the functions, so the names are defined. OK, if we compile with --cudart none then we get link errors because the cuda runtime header doesn't forward declare with __attribute((weak)) but let me sweep that under the carpet.

We can compile against 11.8 but deploy RAPIDS to a machine that has only 11.1 on it. In those cases we need to gracefully fallback at runtime instead of failing to load symbols. That is why we mark the symbols as weak, so we don't get the loader errors when the symbols aren't found with a runtime 11.1

Reply:
Ah thanks, now I understand